### PR TITLE
Add asc/desc and compare support to Enum.sort

### PR DESCRIPTION
**Pull request by josevalim**
Before this PR, sorting dates, decimals, and other structs was
somewhat awkward:

    Enum.sort(dates, &(Date.compare(&1, &2) != :lt)) # ascending
    Enum.sort(dates, &(Date.compare(&1, &2) != :gt)) # descending

Sorting a user by dates, decimals, and other structs was equally
verbose:

    Enum.sort_by(users, &(&1.birthday), &(Date.compare(&1, &2) != :lt)) # ascending
    Enum.sort_by(users, &(&1.birthday), &(Date.compare(&1, &2) != :gt)) # descending

Even a general sort in reverse required passing &>=/2 as an argument,
which is not clear in intent:

    Enum.sort(list, &>=/2) # descending

This PR addresses this by:

  1. Allow `:asc` and `:desc` to be given as sorting functions.

  2. Allow a module to be given to sort, which will automatically
    compare/2. The module can also be wrapped in a :asc/:desc tuple.

All of the examples above can be written as:

    Enum.sort(dates, Date) # ascending
    Enum.sort(dates, {:desc, Date}) # descending
    Enum.sort_by(users, &(&1.birthday), Date) # ascending
    Enum.sort_by(users, &(&1.birthday), {:desc, Date}) # descending
    Enum.sort(list, :desc) # descending

### CODE REVIEW

**Comment by devonestes** [comment]
For those wondering, I benchmarked the difference between `enumerable |> Enum.sort() |> Enum.reverse()` and `Enum.sort(enumerable, :desc)`, and it's a nice improvement both in time and space!

```
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz
Number of Available Cores: 8
Available memory: 15.39 GB
Elixir 1.10.0-dev
Erlang 21.2.6

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 1 s
parallel: 1
inputs: large, medium, small
Estimated total run time: 1.30 min


Benchmarking sort desc with input large...
Benchmarking sort desc with input medium...
Benchmarking sort desc with input small...
Benchmarking sort reverse with input large...
Benchmarking sort reverse with input medium...
Benchmarking sort reverse with input small...

##### With input small #####
Name                   ips        average  deviation         median         99th %
sort desc         192.49 K        5.19 μs   ±375.72%        4.85 μs        6.45 μs
sort reverse      170.03 K        5.88 μs   ±298.83%        5.49 μs        8.60 μs

Comparison: 
sort desc         192.49 K
sort reverse      170.03 K - 1.13x slower

Memory usage statistics:

Name            Memory usage
sort desc            7.13 KB
sort reverse         9.18 KB - 1.29x memory usage

**All measurements for memory usage were the same**

##### With input medium #####
Name                   ips        average  deviation         median         99th %
sort desc          20.23 K       49.42 μs     ±8.87%       47.94 μs       56.43 μs
sort reverse       17.63 K       56.72 μs    ±11.14%       55.31 μs       85.76 μs

Comparison: 
sort desc          20.23 K
sort reverse       17.63 K - 1.15x slower

Memory usage statistics:

Name            Memory usage
sort desc           70.41 KB
sort reverse        88.70 KB - 1.26x memory usage

**All measurements for memory usage were the same**

##### With input large #####
Name                   ips        average  deviation         median         99th %
sort desc           1.68 K      593.79 μs    ±13.44%      559.59 μs      888.22 μs
sort reverse        1.36 K      732.86 μs    ±18.85%      686.55 μs     1231.65 μs

Comparison: 
sort desc           1.68 K
sort reverse        1.36 K - 1.23x slower

Memory usage statistics:

Name            Memory usage
sort desc          703.23 KB
sort reverse       878.10 KB - 1.25x memory usage

**All measurements for memory usage were the same**
```

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -2278,7 +2278,8 @@ defmodule Enum do
   Sorts the `enumerable` by the given function.
 
   This function uses the merge sort algorithm. The given function should compare
-  two arguments, and return `true` if the first argument precedes the second one.
+  two arguments, and return `true` if the first argument is **less than or equal to**
+  the second one.
```
Comment (ericmj) [inline lib/elixir/lib/enum.ex]: I think the previous wording is better you are not necessarily sorting based less / greater than. How about this instead?

The given function should compare two arguments, and return `true` if the first argument precedes or is in the same place as the second one.

File: lib/elixir/lib/enum.ex [hunk 2]
```diff
@@ -2361,29 +2389,79 @@ defmodule Enum do
...
+
+      iex> Enum.sort_by(["some", "kind", "of", "monster"], &byte_size/1, &>=/2)
+      ["monster", "some", "kind", "of"]
+
+  Or use `:asc` and `:desc`:
+
+      iex> Enum.sort_by(["some", "kind", "of", "monster"], &byte_size/1, :desc)
+      ["monster", "some", "kind", "of"]
+
+  As in `sort/2`, avoid using the default sorted to sort structs, as by default
```
Comment (ericmj) [inline lib/elixir/lib/enum.ex]:
```suggestion
  As in `sort/2`, avoid using the default sorting function to sort structs, as by default
```

File: lib/elixir/lib/enum.ex [hunk 3]
```diff
@@ -2316,43 +2327,60 @@ defmodule Enum do
 
   For this reason, most structs provide a "compare" function, such as
   `Date.compare/2`, which receives two structs and returns `:lt` (less than),
-  `:eq` (equal), and `:gt` (greather than). For example, to sort dates
-  increasingly, one would do:
+  `:eq` (equal), and `:gt` (greather than). If you pass a module as the
+  sorting function, Elixir will automatically use the `compare` function
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  sorting function, Elixir will automatically use the `compare/2` function
```

File: lib/elixir/lib/enum.ex [hunk 4]
```diff
@@ -2316,43 +2327,60 @@ defmodule Enum do
...
+  of said module:
 
       iex> dates = [~D[2019-01-01], ~D[2020-03-02], ~D[2019-06-06]]
-      iex> Enum.sort(dates, &(Date.compare(&1, &2) != :gt))
+      iex> Enum.sort(dates, Date)
       [~D[2019-01-01], ~D[2019-06-06], ~D[2020-03-02]]
 
-  Or in decreasing order:
+  To retrieve all dates in descending order, you can wrap the module in
+  a tuple with `:asc` or `:desc` as first argument:
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  a tuple with `:asc` or `:desc` as first element:
```

File: lib/elixir/lib/enum.ex [hunk 5]
```diff
@@ -2316,43 +2327,60 @@ defmodule Enum do
...
+  the mapped elements using the `sorter` function, which defaults
+  to `Kernel.<=/2`.
 
   `sort_by/3` differs from `sort/2` in that it only calculates the
   comparison value for each element in the enumerable once instead of
-  once for each element in each comparison.
-  If the same function is being called on both elements, it's also more
-  compact to use `sort_by/3`.
+  once for each element in each comparison. If the same function is
+  being called on both elements, it's more effiicient to use `sort_by/3`.
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  being called on both elements, it's more efficient to use `sort_by/3`.
```

File: lib/elixir/lib/enum.ex [hunk 6]
```diff
@@ -2361,29 +2389,79 @@ defmodule Enum do
...
+
+  Or use `:asc` and `:desc`:
+
+      iex> Enum.sort_by(["some", "kind", "of", "monster"], &byte_size/1, :desc)
+      ["monster", "some", "kind", "of"]
+
+  As in `sort/2`, avoid using the default sorted to sort structs, as by default
+  it performs structural comparison instead of a semantic one. In such cases,
+  you shall pass a sorting function as third element or any module that implements
+  a `compare` function. For example, to sort users by their birthday in both
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  a `compare/2` function. For example, to sort users by their birthday in both
```

File: lib/elixir/lib/enum.ex [hunk 7]
```diff
@@ -2361,29 +2389,79 @@ defmodule Enum do
...
+  ascending and descending order respectivelly:
```
Comment (fertapric) [inline lib/elixir/lib/enum.ex]:
```suggestion
  ascending and descending order respectively:
```

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**